### DISCUSSION TITLE
Allow loading font files from node_modules

### DISCUSSION THREAD

**Discussion by phuoccss** [remix-run/remix#5200] [Proposals]
### What version of Remix are you using?

1.7.6

### Steps to Reproduce

Use Remix starter js template.
![image](https://user-images.githubusercontent.com/15166715/203100127-d0d54376-36a6-4425-bc38-d276d611477a.png)

Install a fontsource font package.
Import a fontsource stylesheet to `root.jsx`. This works as expected.
Import a font file from the fontsource `node_modules`. esbuild throws an error.
Copy the font file to `app` folder and import it. There's no error.

```js
const karlaCSS = require("@fontsource/karla/400.css");

// this works
const karlaFontFromApp = require("~/karla-latin-400-normal.woff2");

// this doesn't
// const karlaFontFromNodeModules = require("@fontsource/karla/files/karla-latin-400-normal.woff2");

export const links = () => {
  return [
    {
      href: karlaFontFromApp,
      as: `font`,
      type: `font/woff2`,
      crossOrigin: `anonymous`,
      rel: `preload`,
    },
    { href: karlaCSS, rel: `stylesheet` },
  ];
};
```

Here's the error.

![image](https://user-images.githubusercontent.com/15166715/203100149-e2b752b5-cf22-4926-b2c3-86ca86fc403b.png)

### Expected Behavior

Import font file from node_modules successfully and get a url.

Perhaps it can even be automatically added for each route.

### Actual Behavior

esbuild throws an error.

**Comment by machour** [comment]
Converting this to a Proposal discussion, as per our [Development Process](https://remix.run/docs/en/v1/pages/contributing#development-process).

**Comment by jamesarosen** [comment]
This would also help with libraries that export an entrypoint that can be `import`ed into the Remix app, but which then make HTTP requests to load subsequent assets -- e.g. based on theme or internationalization preferences.

TradingView's ChartingLibrary and [Shiki](https://github.com/shikijs/shiki) both do this.

Some nice-to-have features:

* The files should be available as if they were included in `public/`
* I can choose which files or directories within a package are publicly exported. I may not want to have to make the whole package publicly accessible from my site.
* remix-dev watches the files and rebuilds automatically, e.g. if I install a new version

**Comment by frontsideair** [comment]
I'm having the exact same issue right now.

I want to take a moment to emphasize the severity of this problem. When I don't add a preload link, the font file is loaded too late, after all of the scripts. I really would like a way to preload it somehow, and I planned to write a blog post on this exact topic.

**Comment by mikestopcontinues** [comment]
There's an added bit of complexity to preloading fonts. Since you want to preload some font variants AND load the font-family CSS, you also need a way to sync up these values. However, when you import `@fontsource/.../*.css`, the urls to all of the fonts are updated to include hash ids. This means that you can't simply import `@fontsource/.../*.woff2`, since the urls won't match.

I suspect this means the optimal solution is to update the CSS imports to include a map of all assets imported with that CSS file. So the following:

```js
import interCss, {assets} from '@fontsource-variable/inter/index.css';

console.log(assets["/files/inter-latin.woff2"]); // => "/public/build/_assets/inter-latin-A1B2C3D4.woff2"
```

**Reply by mikestopcontinues** [reply]
I should add, in the meantime, I think the best bet is to copy `/node_modules/@fontsource` to `/public` during NPM "postinstall". This isn't ideal, but it's likely the best way to get the desired behavior.

**Reply by frontsideair** [reply]
Your proposal makes a lot of sense, and I believe given the constraints your workaround is the only way for now.
